## Introduction
The python programming language natively implements a number of data structures. Lists, tuples, sets, dictionaries are but some of them. We will be looking at the dictionary data type in subsequent sections.
### What are dictionaries ?
![dictionary image](dictionaries.jpg)

A dictionary in python is a mapping object that maps keys to values, where the keys are unique within a collection and the values can hold any arbitrary value. In addition to being unique, keys are also required to be hashable. An object is said to be hashable if it has a hash value (implemented by a `__hash__()` method) that does not change during the object’s lifetime. Most commonly, we use immutable data types such as strings, integers, and tuples (only if they contain similarly immutable types) as dictionary keys. A dictionary’s data is always enclosed by a pair of curly braces `{ }`.
Typically dictionaries look like this:

In [1]:
my_dict = {"first_name":"John","last_name":"Snow","age":16,"gender":"Male"}

We have created a dictionary named `my_dict` where each key-value pair is separated by a full colon, with the key-value pairs as:
- `first_name` - `John`
- `last_name` - `Snow`
- `age` - `16`
- `gender` - `Male`

### Comparisons
Dictionaries are an implementation of [Associative Arrays](https://en.wikipedia.org/wiki/Associative_array). All Associative arrays have a structure of (key, value) pairs, where each key is unique for every collection. Other languages also have similar executions, such as:
- Maps in Go
- std::map in C++
- Maps in Java
- JavaScript objects

Unlike sequenced data types like lists and tuples, where indexing is achieved using positional indices, dictionaries are indexed using their keys. Therefore, individual values can be accessed using these keys. 
Typically dictionaries store associative data, i.e data that is related. Examples of such data include the attributes of an object, `SQL` query results and csv-formatted information. Throughout this article, we will be using dictionaries to store job listing details from [Kaggle](https://www.kaggle.com/ardenn/brighter-monday-job-listings/data).

## Dictionary Operations

### Creation
We initialize an empty dictionary using a pair of curly braces. This approach is often used when we expect to store some data at later stages of our operation.


In [2]:
empty_dict = {}

In the line above, we have created an empty dictionary named `empty_dict`.

For instances when we have our data beforehand, we use curly braces with the key-value pairs. We can now create a dictionary to represent the second row of data in the `jobs.csv` file.

In [3]:
job1 = {"title":"Production Manager","location":"Rest of Kenya","job_type":"Full Time",
             "employer":"The African Talent Company (TATC)","category":"Farming"}

We just created a dictionary with the keys `title`,` location`, `job_type`, `employer`, `category` and assigned it to the variable `job1`.

Dictionaries can also be created using the `dict()` constructor. To do this we pass the constructor a sequence of key-value pairs. We could also pass in named arguments. Let's create a dictionary to represent the third row of data in the `jobs.csv` file, using both of these methods.

In [4]:
# create an empty dictionary
empty_property = dict()

# create dictionary using a list of key-value tuples
job2 = dict([("title","Marketing & Business Development Manager"),("location","Mombasa"),\
("job_type","Full Time"),\
("employer","KUSCCO Limited (Kenya Union of Savings & Credit Co-operatives Limited)"),\
("category","Marketing & Communications")])

We passed a sequence, in this case a list of key-value tuples, to the `dict()` constructor to create our dictionary, and assigned it to the variable `job2`.

In [5]:
# Using keyword arguments
dict(title="Marketing & Business Development Manager",location="Mombasa",job_type="Full Time",
     employer="KUSCCO Limited (Kenya Union of Savings & Credit Co-operatives Limited)",
     category="Marketing & Communications")

{'title': 'Marketing & Business Development Manager',
 'location': 'Mombasa',
 'job_type': 'Full Time',
 'employer': 'KUSCCO Limited (Kenya Union of Savings & Credit Co-operatives Limited)',
 'category': 'Marketing & Communications'}

Here, we created a dictionary using named arguments. The keys are the argument names, while the values are the argument values. It is however important to note that this method is only suitable when our keys are just simple strings.

### Accessing Items
As we mentioned earlier on, dictionaries are indexed using their keys.To access a particular value in a dictionary we use the indexing operator (key inside square brackets). However, to use this method, we need to make sure the key we intend to retrieve exists, lest we get a `KeyError`. Checking for availability of a key is as easy as using the `in` operator.

If you are like me, this is probably a lot of work. The good news, however, is that we have a better tool - the `get()` method. This method works by giving out a value if the key exists, or returning `None`. `None` sounds better than an error, right? What if we want to go even further, and return something, a placeholder of sorts? `get()` takes a second argument, a default value to be used in place of `None`. Now Let's use `in` to check if `title` exists in `job2`, then we can use indexing to retrieve its value. We'll also go ahead and use `get()` to retrieve `salary` from `job2`.

In [6]:
# Check existence of title
"title" in job2 # returns True
"salary" in job2 # returns False

# Using key indexing
job2["title"] # return 'Marketing & Business Development Manager'

# Using get() method
job2.get("title") # return 'Marketing & Business Development Manager'
job2.get("salary") # return None

# Passing a second argument to get()
job2.get("salary", 5000) #return 5000

5000

In the example above we use indexing to access the `title` from `job2` after making sure it is available using `in`. Similarly, we use `get()` to access the `title` and `salary`.  

However, `job2` doesn't have a `salary` key so the return value is `None`. Adding a second argument, to `get()` now gives us `5000` instead of `None`.

### Modification
Dictionaries can be modified directly using the keys or using the `update()` method. `update()` takes in a dictionary with the key-value pairs to be modified or added. For our demonstration, let's:
- Add a new item (salary) to `job2` with a value of 10000.
- Modify the `job_type` to be "Part time".
- Update the `salary` to 20000.
- Update the dictionary to include a new item (available) with a value of `True`.

In [7]:
# Adding a new entry for salary using the index
job2["salary"] = 10000

# Modifying the entry for job_type using the index
job2["job_type"] = "Part time"

# Modifying the salary entry using update
job2.update({"salary":20000})

# Adding the available entry using update
job2.update({"available":True})

To add a new entry, we use syntax similar to indexing. If the key exists, then the value will be modified, however, if the key doesn't exist, a new entry will be created with the specified key and value. 
Initially, we assigned a value of 10000 to the `salary` key, but since `salary` doesn't exist, a new entry is created, with that value. For our second example, the `job_type` key exists, the value is modified to "Part time". Next, we use the `update()` method to change the `salary` value to 20000, since `salary` is already a key in the dictionary. Finally, we apply `update()` to the dictionary, a new entry is created with a key of `available` and value of `True`.

### Deletion
 We can now remove the just created `salary` entry from `job2`, and remove everything from `job1`.

In [8]:
del job2["salary"]
del job2["available"]
print(job2) #return a dictionary without 'salary' and 'available' entries

job1.clear()
print(job1) #return an empty dictionary

del job1

{'title': 'Marketing & Business Development Manager', 'location': 'Mombasa', 'job_type': 'Part time', 'employer': 'KUSCCO Limited (Kenya Union of Savings & Credit Co-operatives Limited)', 'category': 'Marketing & Communications'}
{}


To remove the entries associated with the `salary` and `available` keys from `job2`, we use the `del` keyword. Now if we go ahead and print `job2`, the `salary` and `available` entries are gone. Removing all items from `job1` entails using the `clear()` method, which leaves us with an empty dictionary. If we don't need a dictionary anymore, say `job1`, we use the `del` keyword to delete it. Now if we try printing `job1` we'll get a `NameError` since `job1` is no longer defined.

### Iteration
Dictionaries are iterable, and we can iterate through them in 3 different ways:
 - `dict.values()` - this returns an iterable of the dictionary's values.
 - `dict.keys()` - this returns an iterable of the dictionary's keys.
 - `dict.items()` - this returns an iterable of the dictionary's (key,value) pairs.

But why would we need to iterate over a dictionary? Our dataset has about 860 listings, suppose we wanted to display the properties of all these on our website, it wouldn't make sense to write the same markup 860 times. It would be efficient to dynamically render the data using a loop.
Let's iterate over `job2` using a `for-loop` using all the three methods. Furthermore we'll use the `csv` module to read our csv-formatted data in to a list of dictionaries, then we'll iterate through all the dictionaries and print out the keys and values.

In [9]:
# Using values()
for val in job2.values():
    print(val) #prints the values of job2
    
# Using keys()
for key in job2.keys():
    print(key) #prints the keys of job2

# Dictionary iteration use case
import csv
with open('jobs.csv','r') as csv_file:
    reader = csv.DictReader(csv_file)
    for job in reader:
        
        # Using items()
        for key,val in job.items():
            # Apply any additional processing
            print(key, val) #print the keys and values of each job

Marketing & Business Development Manager
Mombasa
Part time
KUSCCO Limited (Kenya Union of Savings & Credit Co-operatives Limited)
Marketing & Communications
title
location
job_type
employer
category
title Production Manager
location Rest of Kenya
job_type Full Time
employer The African Talent Company (TATC)
category Farming
title Marketing & Business Development Manager
location Mombasa
job_type Full Time
employer KUSCCO Limited (Kenya Union of Savings & Credit Co-operatives Limited)
category Marketing & Communications
title Loans Manager
location Mombasa
job_type Full Time
employer KUSCCO Limited (Kenya Union of Savings & Credit Co-operatives Limited)
category Accounting & Auditing
title Internal Auditor
location Mombasa
job_type Full Time
employer KUSCCO Limited (Kenya Union of Savings & Credit Co-operatives Limited)
category Accounting & Auditing
title Accounts Clerk
location Rest of Kenya
job_type Full Time
employer Machakos Institute Of Technology
category Accounting & Auditing
ti

location Nairobi
job_type Full Time
employer Well Told Story
category Marketing & Communications
title Backend Developer
location Nairobi
job_type Full Time
employer Well Told Story
category IT & Telecoms
title Account Manager (Retail Merchandising – Instore)
location Nairobi
job_type Full Time
employer TDF Advertising | The Dream Factory
category Sales & Business Development
title Dispatch Assistant
location Mombasa
job_type Full Time
employer Aga Khan Hospital, Mombasa
category Accounting & Auditing
title Patient Services Cashier / Receptionist – Ohc, Nyali Branch
location Mombasa
job_type Full Time
employer Aga Khan Hospital, Mombasa
category Administrative
title Accounts Assistant
location Mombasa
job_type Full Time
employer Aga Khan Hospital, Mombasa
category Accounting & Auditing
title Network Support Intern
location Kisumu
job_type Internships & Graduate
employer Uwezo Broadband Limited
category IT & Telecoms
title Technician
location Kisumu
job_type Full Time
employer Uwezo Bro

location Outside Kenya
job_type Full Time
employer Chemonics International Inc
category Project Management
title Senior Regional Advisor
location Nairobi
job_type Full Time
employer US Agency for International Development
category Project Management
title Members
location Rest of Kenya
job_type Full Time
employer Uasin Gishu County Public Service Board
category Accounting & Auditing
title Chairperson
location Rest of Kenya
job_type Full Time
employer Uasin Gishu County Public Service Board
category Accounting & Auditing
title Members - Alcoholic Drinks Control Board
location Rest of Kenya
job_type Full Time
employer Uasin Gishu County Public Service Board
category Management
title Director Of Criminal Investigation
location Nairobi
job_type Full Time
employer National Police Service Commission - NPSC
category Management
title Deputy Inspectors General Of Police
location Nairobi
job_type Full Time
employer National Police Service Commission - NPSC
category Management
title Project Super

title Receptionist
location Nairobi
job_type Full Time
employer Right Corporation (Shenzhen Right Net Tech Co., Ltd)
category Administrative
title Senior Human Resource And Administration Officer 
location Nairobi
job_type Full Time
employer MINISO LIFESTYLE KENYA LIMITED
category Recruitment
title Auto Electrical Technician 
location Nairobi
job_type Full Time
employer Dashcam Center
category Engineering
title Intern Portfolio Quality Officer
location Mombasa
job_type Internships & Graduate
employer Watu Credit Limited
category Administrative
title Finance Officer
location Nairobi
job_type Full Time
employer The World Bank Group (WBG)
category Accounting & Auditing
title Finance Analyst
location Nairobi
job_type Full Time
employer The World Bank Group (WBG)
category Accounting & Auditing
title Microfinance Risk Officer
location Nairobi
job_type Full Time
employer Corporate Staffing Services Ltd
category Quality Control & Assurance
title Loan Officer
location Nairobi
job_type Full Time

title Sales Executives
location Mombasa
job_type Full Time
employer G. North & Son Limited
category Sales & Business Development
title Chief Research And Development Officer
location Nairobi
job_type Full Time
employer Kenya Veterinary Vaccines Production Institute
category Medicine & Pharmaceutical
title Deputy Director, Technical Services
location Nairobi
job_type Full Time
employer Kenya Veterinary Vaccines Production Institute
category Medicine & Pharmaceutical
title Executive Assistant/administration Assistant
location Nairobi
job_type Full Time
employer The African Talent Company (TATC)
category Administrative
title Finance & Administration Manager
location Nairobi
job_type Full Time
employer The African Talent Company (TATC)
category Accounting & Auditing
title Clerk Of Works (Engineering)
location Nairobi
job_type Full Time
employer The African Talent Company (TATC)
category Engineering
title Research Intern
location Nairobi
job_type Internships & Graduate
employer Consumer Ins

category Administrative
title Marketing Officer
location Nairobi
job_type Full Time
employer International Medical Treatment (IMT)
category Marketing & Communications
title Administrative Assistant (Office Manager)
location Nairobi
job_type Full Time
employer Skills Geographic
category Consulting & Strategy
title Community Manager
location Nairobi
job_type Full Time
employer Salim Wazaran Kenya Co. Ltd
category Marketing & Communications
title Assistant Internal Auditor
location Nairobi
job_type Full Time
employer FSI Capital Ltd
category Accounting & Auditing
title Accounts Assistant
location Nairobi
job_type Full Time
employer Hand in Hand Eastern Africa (HiH EA)
category Accounting & Auditing
title Creative Designer
location Nairobi
job_type Full Time
employer AutoXpress Ltd
category Art & Design
title Legal Assistant
location Nairobi
job_type Full Time
employer Anonymous
category Legal
title Credit Control Manager - Tanzania
location Outside Kenya
job_type Full Time
employer Uber A

job_type Full Time
employer Public Service Commission
category Medicine & Pharmaceutical
title Senior Medical Specialist – Four (4) Posts (Obstetrics/gynecologist – 1, Paeditrician -1, Ophthalmologist -1, Dermatologist -1)
location Nairobi
job_type Full Time
employer Public Service Commission
category Medicine & Pharmaceutical
title Senior Assistant Director, Nursing Services 
location Nairobi
job_type Full Time
employer Public Service Commission
category Medicine & Pharmaceutical
title Deputy Director, Nursing Services
location Nairobi
job_type Full Time
employer Public Service Commission
category Medicine & Pharmaceutical
title Senior Assistant Chief Radiation Protection Officer 
location Nairobi
job_type Full Time
employer Public Service Commission
category Medicine & Pharmaceutical
title Deputy Chief Radiation Protection Officer 
location Nairobi
job_type Full Time
employer Public Service Commission
category Medicine & Pharmaceutical
title Chief Radiation Protection Officer 
locati

category Project Management
title Kiswahili Teacher (High School) - Machakos
location Rest of Kenya
job_type Full Time
employer Springvale High School
category Teaching & Training
title Medical Sales Representative
location Nairobi
job_type Full Time
employer Anonymous
category Sales & Business Development
title Contractor - Accounts Payables
location Nairobi
job_type Full Time
employer Safaricom Limited
category Accounting & Auditing
title Risk & Compliance Manager
location Nairobi
job_type Full Time
employer The African Talent Company (TATC)
category Quality Control & Assurance
title Chef - Thika
location Thika
job_type Full Time
employer Anonymous
category Food Services
title Food & Beverage Supervisor
location Thika
job_type Full Time
employer Anonymous
category Food Services
title Ecde And P1 Teacher - Ongata Rongai
location Rest of Kenya
job_type Full Time
employer The Azuri School
category Teaching & Training
title Copywriter And Writer - Uganda
location Outside Kenya
job_type F

category Food Services
title Accountant
location Nairobi
job_type Full Time
employer Westlands Medical Centre
category Accounting & Auditing
title Business Relationship Officer
location Nairobi
job_type Full Time
employer Springboard Capital Ltd
category Sales & Business Development
title Finance Manager
location Nairobi
job_type Full Time
employer Riley Services Ltd
category Accounting & Auditing
title Sales And Marketing Executive
location Nairobi
job_type Full Time
employer Hotman Trading Limited
category Sales & Business Development
title Sales And Marketing Executive
location Nairobi
job_type Full Time
employer Anonymous
category Sales & Business Development
title Retail Sales Executive 
location Nairobi
job_type Full Time
employer Kenbro Industries Limited
category Sales & Business Development
title Heavy Diesel Mechanics Wanted In New Zealand
location Outside Kenya
job_type Full Time
employer GoughGough & Hamer
category Trades & Services
title Country Financial Controller
locati

employer Dilla University
category Teaching & Training
title Lecturer/msc - Pharmacognosy
location Outside Kenya
job_type Full Time
employer Dilla University
category Teaching & Training
title Asst. Professor - Phd In Public Health 
location Outside Kenya
job_type Full Time
employer Dilla University
category Teaching & Training
title Asst. Professor - Phd In Reproductive Health 
location Outside Kenya
job_type Full Time
employer Dilla University
category Teaching & Training
title Asst. Professor - Phd In Mental Health 
location Outside Kenya
job_type Full Time
employer Dilla University
category Teaching & Training
title Asst. Professor - Anesthesiologist
location Outside Kenya
job_type Full Time
employer Dilla University
category Teaching & Training
title Assistant Professor And Above - Phd In Numerical Analysis
location Outside Kenya
job_type Full Time
employer Dilla University
category Teaching & Training
title Assistant Professor And Above - Phd In Algebra
location Outside Kenya
job

First, we loop through the iterable `job2.values()` and print out the value during each iteration. Secondly, we iterate through `job2.keys()` while printing out the key. This is the default behaviour when iterating over a dictionary without specifying whether we want the keys, values or items. Finally, we step through the list of dictionaries, and for each one, loop through the keys and values simultaneously. We include both key and value in the for-loop constructor since `job.items()` yields a tuple of key and value during each iteration. We can now apply any kind of operation to our data at this point. Our implementation simply prints out the pair at each step.

### Sorting
Borrowing from our description of dictionaries earlier, this data type in meant to be unordered, and doesn't come with the sorting functionality. Calling the `sorted()` function and passing it a dictionary only returns a list of the keys in a sorted order.

If we use the `items()` iterable we could sort the items of our dictionary as we please. However, this doesn't give us our original dictionary, but an array of key-value tuples in a sorted order. Say we wanted to display the job details in the above example in alphabetical order, We would need to alter our iteration to give sorted results. Lets walk through the example again an see how we would achieve that functionality.

In [10]:
with open('jobs.csv','r') as csv_file:
    reader = csv.DictReader(csv_file)
    for job in reader:
        
        # Using sorted() to sort a dictionary's items on the keys
        for key,val in sorted(job.items(),key=lambda item:item[0]):
        
            # Apply any additional processing
            print(key, val) #print the keys and values of each job


category Farming
employer The African Talent Company (TATC)
job_type Full Time
location Rest of Kenya
title Production Manager
category Marketing & Communications
employer KUSCCO Limited (Kenya Union of Savings & Credit Co-operatives Limited)
job_type Full Time
location Mombasa
title Marketing & Business Development Manager
category Accounting & Auditing
employer KUSCCO Limited (Kenya Union of Savings & Credit Co-operatives Limited)
job_type Full Time
location Mombasa
title Loans Manager
category Accounting & Auditing
employer KUSCCO Limited (Kenya Union of Savings & Credit Co-operatives Limited)
job_type Full Time
location Mombasa
title Internal Auditor
category Accounting & Auditing
employer Machakos Institute Of Technology
job_type Full Time
location Rest of Kenya
title Accounts Clerk
category Marketing & Communications
employer Machakos Institute Of Technology
job_type Full Time
location Rest of Kenya
title Marketing Officers
category Teaching & Training
employer Machakos Institute

category Food Services
employer Fairview Hotel and Town Lodge by City Lodge Hotel Group
job_type Full Time
location Nairobi
title Executive Chef
category Marketing & Communications
employer Well Told Story
job_type Full Time
location Nairobi
title Events And Production
category IT & Telecoms
employer Well Told Story
job_type Full Time
location Nairobi
title Backend Developer
category Sales & Business Development
employer TDF Advertising | The Dream Factory
job_type Full Time
location Nairobi
title Account Manager (Retail Merchandising – Instore)
category Accounting & Auditing
employer Aga Khan Hospital, Mombasa
job_type Full Time
location Mombasa
title Dispatch Assistant
category Administrative
employer Aga Khan Hospital, Mombasa
job_type Full Time
location Mombasa
title Patient Services Cashier / Receptionist – Ohc, Nyali Branch
category Accounting & Auditing
employer Aga Khan Hospital, Mombasa
job_type Full Time
location Mombasa
title Accounts Assistant
category IT & Telecoms
employe

category Teaching & Training
employer Janta Kenya Limited
job_type Full Time
location Nairobi
title Early Childhood Development Education Instructor / Trainer
category Teaching & Training
employer Janta Kenya Limited
job_type Full Time
location Nairobi
title Human Resource Management Instructor / Trainer
category Teaching & Training
employer Janta Kenya Limited
job_type Full Time
location Nairobi
title Hospitality & Tourism Management Instructor / Trainer
category Administrative
employer Janta Kenya Limited
job_type Full Time
location Nairobi
title Receptionist / Office Administrator
category Property Management
employer Janta Kenya Limited
job_type Full Time
location Nairobi
title Caterer 
category IT & Telecoms
employer PowerGen Renewable Energy Ltd
job_type Full Time
location Nairobi
title Business Systems – Erp Expert
category IT & Telecoms
employer PowerGen Renewable Energy Ltd
job_type Full Time
location Nairobi
title Business Systems – Implementation Consultant
category IT & Tel

title Program Development And Quality Coordinator
category Consulting & Strategy
employer Bureau for Institutional Reform and Democracy
job_type Full Time
location Nairobi
title Long-Term Key Expert (Indicatively Full Time 57 Months) – Legal Aid Expert
category Consulting & Strategy
employer Bureau for Institutional Reform and Democracy
job_type Full Time
location Nairobi
title Non-Key Expert (Part-Time) – Monitoring And Evaluation Expert
category Consulting & Strategy
employer Bureau for Institutional Reform and Democracy
job_type Full Time
location Nairobi
title Non-Key Expert (Part-Time) – Regulatory Drafting Expert (Legal Aid)
category Project Management
employer International Organization for Migration (IOM)
job_type Full Time
location Nairobi
title Regional Emergency And Post-Crisis Officer
category Supply Chain & Procurement
employer International Organization for Migration (IOM)
job_type Full Time
location Nairobi
title Compliance Officer
category Accounting & Auditing
employer

title Sales Account Manager
category Administrative
employer The African Talent Company (TATC)
job_type Full Time
location Nairobi
title Receptionist/ Customer Service 
category Administrative
employer Right Corporation (Shenzhen Right Net Tech Co., Ltd)
job_type Full Time
location Nairobi
title Receptionist
category Recruitment
employer MINISO LIFESTYLE KENYA LIMITED
job_type Full Time
location Nairobi
title Senior Human Resource And Administration Officer 
category Engineering
employer Dashcam Center
job_type Full Time
location Nairobi
title Auto Electrical Technician 
category Administrative
employer Watu Credit Limited
job_type Internships & Graduate
location Mombasa
title Intern Portfolio Quality Officer
category Accounting & Auditing
employer The World Bank Group (WBG)
job_type Full Time
location Nairobi
title Finance Officer
category Accounting & Auditing
employer The World Bank Group (WBG)
job_type Full Time
location Nairobi
title Finance Analyst
category Quality Control & Assu

title Technical Sales Personnel (Energy Audits)
category Sales & Business Development
employer Lean Energy Solutions Ltd (Lean Solutions Group)
job_type Internships & Graduate
location Nairobi
title Paid Sales Interns
category Sales & Business Development
employer G. North & Son Limited
job_type Full Time
location Mombasa
title Sales Executives
category Medicine & Pharmaceutical
employer Kenya Veterinary Vaccines Production Institute
job_type Full Time
location Nairobi
title Chief Research And Development Officer
category Medicine & Pharmaceutical
employer Kenya Veterinary Vaccines Production Institute
job_type Full Time
location Nairobi
title Deputy Director, Technical Services
category Administrative
employer The African Talent Company (TATC)
job_type Full Time
location Nairobi
title Executive Assistant/administration Assistant
category Accounting & Auditing
employer The African Talent Company (TATC)
job_type Full Time
location Nairobi
title Finance & Administration Manager
category 

location Nairobi
title Mobile Ux/ui Engineer
category Marketing & Communications
employer Salvon Africa Ltd
job_type Full Time
location Nairobi
title Online/digital Marketing
category Supply Chain & Procurement
employer Taimba Ltd
job_type Full Time
location Nairobi
title Warehouse Clerk - Entry Level
category Teaching & Training
employer Midas International Boys High School (Midas Group of Schools)
job_type Full Time
location Thika
title English/literature Teacher
category Management
employer Public Service Commission
job_type Full Time
location Nairobi
title Deputy Director, Irrigation Water Management
category Management
employer Public Service Commission
job_type Full Time
location Nairobi
title Director, Irrigation Water Management 
category Management
employer Public Service Commission
job_type Full Time
location Nairobi
title Principal Superintending Geologist
category Management
employer Public Service Commission
job_type Full Time
location Nairobi
title Senior Principal Superi

location Nairobi
title Deputy County Commissioner I 
category Marketing & Communications
employer Anonymous
job_type Full Time
location Rest of Kenya
title Sales Executives
category Administrative
employer Public Service Commission
job_type Full Time
location Nairobi
title Deputy Secretary 
category Administrative
employer Public Service Commission
job_type Full Time
location Nairobi
title Deputy Secretary 
category Administrative
employer Public Service Commission
job_type Full Time
location Nairobi
title Senior Deputy Secretary
category Administrative
employer Public Service Commission
job_type Full Time
location Nairobi
title Director Of Administration
category Project Management
employer Church of Sweden
job_type Full Time
location Nairobi
title Senior Coordinator For The East Africa Refugee Response
category Project Management
employer International Committee of the Red Cross (ICRC)
job_type Full Time
location Nairobi
title Regional Stress & Resilience Advisor - Based In Nairobi (

location Nairobi
title Cyber Café Administrator
category Trades & Services
employer Villa Rosa Kempinski
job_type Full Time
location Nairobi
title Spa Therapist
category Quality Control & Assurance
employer East African Breweries Limited (EABL)
job_type Full Time
location Nairobi
title Quality Analyst
category IT & Telecoms
employer Open Capital Advisors
job_type Full Time
location Nairobi
title Systems Specialist
category Management
employer Open Capital Advisors
job_type Full Time
location Nairobi
title Project Leader
category Accounting & Auditing
employer Barclays Bank of Kenya Limited
job_type Full Time
location Nairobi
title Internal Auditor - Africa Subsidiaries It
category IT & Telecoms
employer Schneider Electric Kenya
job_type Full Time
location Nairobi
title It Division Enterprise & Systems Account Manager 
category Customer Service & Support
employer Afrikaloan.com
job_type Full Time
location Nairobi
title Call Centre Operator - Fintech Company
category Customer Service & S

location Outside Kenya
title Lecturer/msc - Pharmaceutics
category Teaching & Training
employer Dilla University
job_type Full Time
location Outside Kenya
title Lecturer/msc - Medicinal Chemistry
category Teaching & Training
employer Dilla University
job_type Full Time
location Outside Kenya
title Lecturer/msc - Pharmacognosy
category Teaching & Training
employer Dilla University
job_type Full Time
location Outside Kenya
title Asst. Professor - Phd In Public Health 
category Teaching & Training
employer Dilla University
job_type Full Time
location Outside Kenya
title Asst. Professor - Phd In Reproductive Health 
category Teaching & Training
employer Dilla University
job_type Full Time
location Outside Kenya
title Asst. Professor - Phd In Mental Health 
category Teaching & Training
employer Dilla University
job_type Full Time
location Outside Kenya
title Asst. Professor - Anesthesiologist
category Teaching & Training
employer Dilla University
job_type Full Time
location Outside Kenya
ti

In this example we use python's inbuilt `sorted()` function which takes in an iterable (our dictionary's items). The `key` argument of the `sorted()` function instructs `sorted()` to use the value at index 0 for sorting. Similarly, to sort by the values, we use index 1 instead of index 0.

### Other Methods
Dictionaries have other methods that could be used on demand. To read up further on these, please consult the [python documentation](https://docs.python.org/3/library/stdtypes.html#typesmapping). Here are some other useful methods:
- `pop(key,default)` - deletes the key `key` and returns it, or returns an optional `default` when the key doesn't exist.
- `len(d)` - returns the number of items in a dictionary `d`.
- `copy()` - returns a shallow copy of the original. This shallow copy has similar references to the original, and not copies of the original's items. 
- `setdefault(key,default)` - returns the value of `key` if in the dictionary, or sets the new key with an optional `default` as its value then returns the value.

## Speeding Up your Code

Dictionary unpacking can greatly speed up our code. It involves destructuring a dictionary into individual keyword arguments with values. This is especially useful for cases that involve supplying multiple keyword arguments for example in function calls. To implement this functionality we use the iterable unpacking operator (`**`). 

What if we needed `Job` objects to work with, instead of dictionaries? We shouldn't have to do some heavy lifting to get our data reorganized in to objects. Let's see how we could translate our dictionaries into objects, by again tweaking our previous code.

In [11]:
from jobs import Job

# Creating a job object without unpacking
Job("Marketing & Business Development Manager","Mombasa","Full Time",\
     "KUSCCO Limited (Kenya Union of Savings & Credit Co-operatives Limited)",\
     "Marketing & Communications")


with open('jobs.csv','r') as csv_file:
    reader = csv.DictReader(csv_file)
    for job in reader:
        
        # Creating a job object with unpacking
        Job(**job2)

To instantiate a new `Job` object, traditionally, we would need to pass in all the required arguments. However, with unpacking, we just pass in a dictionary with the `**` operator before it. The operator unpacks the dictionary in to an arbitrary number of named arguments. This approach is much cleaner and involves less code.

## Antipatterns: Wrong usage
Compared to lists and tuples, dictionaries take up more space in memory, since they need to store both the key and value, as opposed to just values. Therefore, dictionaries should only be used in cases where we have associative data, that would lose meaning if stored in lists. Since dictionaries are unordered, it would not be sensible to store strictly arranged data in them. They are also mutable, and not suitable for storing data than shouldn't be modified in place.

Dictionaries are well-designed to let us find a value instantly without necessarily having to search through the entire collection, hence we should not use loops for such an operation.

In [12]:
# How not to search for a value and return it
key_i_need = "location"
target = ""
for key in job2:
    if key == key_i_need:
        target = job2[key]
        
# How to search efficiently
target = job2.get("location")

We have a variable `key_i_need` containing the key we want to search for. We have used a for loop to traverse the collection, comparing the key at each step with our variable. If we get a match, we assign that key's value to the variable `target`. This is the wrong approach. We should instead use `get()`, and pass it the desired key.

## Performance Tradeoffs
In this section, we will be comparing the space-time trade-offs between dictionaries and objects. We will use the `timeit` module and `getsizeof` function to determine the time and space aspects. The operations we will be testing are:
- Accessing an entry in a dictionary versus accessing a field in a object.
- Adding a new entry to a dictionary versus adding a new field to an object.

In terms of space, we will be testing the following operations:
- Size of object classes.
- Size of the default object.
- Size of an empty dictionary.
- Size of dictionary and object after adding one entry and one field respectively.

Let's start by comparing time differences between accessing a single entry from a dictionary and reading a field from an object.

In [13]:
import timeit
# Accessing entry using indexing operator
timeit.timeit('string="Random String"+job3["employer"]',setup='job3 = {"title":"Loans Manager","location":"Mombasa",\
"job_type":"Full Time","employer":"KUSCCO Limited (Kenya Union of Savings & Credit Co-operatives Limited)",\
"category":"Accounting & Auditing"}',number=10000)

# Accessing entry using get() method
timeit.timeit('string="Random String"+job3.get("employer")',setup='job3 = {"title":"Loans Manager","location":"Mombasa",\
"job_type":"Full Time","employer":"KUSCCO Limited (Kenya Union of Savings & Credit Co-operatives Limited)",\
"category":"Accounting & Auditing"}',number=10000)

# Accessing object field using namespace operator
timeit.timeit('string="Random String"+job3.employer',setup='from jobs import Job; job3 = Job("Loans Manager","Mombasa","Full Time",\
"KUSCCO Limited (Kenya Union of Savings & Credit Co-operatives Limited)",\
"Accounting & Auditing")',number=10000)

0.003080508999119047

For the three tests, we look at the time it takes for each of the operations to execute 10000 loops. From our results, accessing an entry from a dictionary using the indexing operator is the fastest, closely followed by reading a field from an object using the namespace operator. Using `get()` to read an entry turn out to be slowest.

Next we'll look at execution times when adding a new dictionary entry and adding a new field to an object.

In [14]:
# Adding a new entry using indexing and assignment operators
timeit.timeit('job3["salary"]=50000',setup='job3 = {"title":"Loans Manager","location":"Mombasa",\
"job_type":"Full Time","employer":"KUSCCO Limited (Kenya Union of Savings & Credit Co-operatives Limited)",\
"category":"Accounting & Auditing"}',number=10000)

# Adding a new entry using update() method
timeit.timeit('job3.update({"salary":50000})',setup='job3 = {"title":"Loans Manager","location":"Mombasa",\
"job_type":"Full Time","employer":"KUSCCO Limited (Kenya Union of Savings & Credit Co-operatives Limited)",\
"category":"Accounting & Auditing"}',number=10000)

# Adding a new field to the object
timeit.timeit('job3.salary = 50000',setup='from jobs import Job; job3 = Job("Loans Manager","Mombasa",\
"Full Time",\"KUSCCO Limited (Kenya Union of Savings & Credit Co-operatives Limited)",\
"Accounting & Auditing")',number=10000)

0.0015931990055833012

![time.png](time.png)
For this round of tests, adding a new entry to a dictionary using the indexing and assignment operators took the lead early. Adding a new field to an object was almost as fast, but came in second while `update()` method came in last. As far as time is concerned, we can conclude that the dictionary operation was faster.

Let's setup a new test environment for the speed tests, by initializing a `Job` object and creating a new dictionary with the same data.

In [15]:
# create a job object
from jobs import Job
job_object = Job("Loans Manager","Mombasa","Full Time",\
        "KUSCCO Limited (Kenya Union of Savings & Credit Co-operatives Limited)",\
        "Accounting & Auditing")

# Create a job dictionary
job_dict = {"title":"Loans Manager","location":"Mombasa",\
"job_type":"Full Time","employer":"KUSCCO Limited (Kenya Union of Savings & Credit Co-operatives Limited)",\
"category":"Accounting & Auditing"}

# show that the internal values are the same
job_dict == job_object.__dict__

True

We will use `getsizeof()` function to measure the size of our objects, and compare them. `getsizeof()` only measures the size of an object, excluding its fields. However, for dictionaries, it includes the size of the hash table.

In [16]:
from sys import getsizeof

# Size of Job class
getsizeof(Job)

# Size of original Job() object
getsizeof(Job())

# Size of job_object
getsizeof(job_object)

# Size of dict object
getsizeof(dict)

# Size of empty dictionary {}
getsizeof({})

# Size of job_dict
getsizeof(job_dict)

240

From our results, we can tell that the `dict` class has a smaller memory footprint (400) compared to the `Job` class (1184). However, `job_object` has a smaller size (56) than `job_dict` (240)
![empty.png](empty.png)

Now we can proceed to add a new entry to the `job_dict`, and create two new fields in `Job`.`Job2` and `Job3` are implementations of the `Job` class with one and two extra fields respectively.

In [17]:
# Size of job_dict with one additional entry using update()
job_dict2 = job_dict.copy()
job_dict2.update({'salary':50000})
getsizeof(job_dict2)

# Size of job_dict with one additional entry using index and assignment operator
job_dict['salary']=50000
getsizeof(job_dict)

368

In [18]:
from jobs import Job2

# Size of Job2 class - with one extra field
getsizeof(Job2)

# Size of Job2 default object - with one extra field
getsizeof(Job2())

56

In [19]:
from jobs import Job3

# Size of Job3 class - with two extra fields
getsizeof(Job3)

# Size of Job3 default object - with two extra fields
getsizeof(Job3())

56

![space.png](space.png)
Our tests point to the fact that objects take up smaller memory spaces than dictionaries, despite the class definitions being rather large. Dictionaries have a much bigger memory fingerprint due to their hash tables which are re-evaluated during every add operation.

## Conclusion
Dictionaries come in very handy for regular python usage. They are suitable for use with unordered data that relies on relations. Caution should however be exercised to ensure we do not use dictionaries in the wrong way and end up slowing down execution of our code. For further reading please refer to the official python documentation on [mapping types](https://docs.python.org/3/library/stdtypes.html#typesmapping).